x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/AutodetectResultProcessor.java

```diff
@@ -288,9 +288,15 @@ void processResult(AutodetectResult result) {
             // Commit previous writes here, effectively continuing
             // the flush from the C++ autodetect process right
             // through to the data store
-            bulkResultsPersister.executeRequest();
-            persister.commitResultWrites(jobId);
-            flushListener.acknowledgeFlush(flushAcknowledgement);
+            try {
+                bulkResultsPersister.executeRequest();
+                persister.commitResultWrites(jobId);
+            } catch (Exception e) {
+                LOGGER.error("[" + jobId + "] failed to bulk persist results and commit writes during flush acknowledgement.", e);
+            } finally {
+                LOGGER.debug("[{}] Flush acknowledgement sent to listener for ID {}", jobId, flushAcknowledgement.getId());
+                flushListener.acknowledgeFlush(flushAcknowledgement);
```

Comment:
If there's been an exception it would be better if the caller received the exception rather than an acknowledgement saying the flush worked fine.

I think this can be achieved by:

1. Making `FlushListener.acknowledgeFlush` take two arguments, namely a `FlushAcknowledgement` and an `Exception`, one of which should be `null`.
2. Making `FlushAcknowledgementHolder` hold an `Exception` as well as a latch and a `FlushAcknowledgement`.
3. Making `FlushListener.waitForFlush` throw the exception in the holder if it's non-`null`.  This will get propagated to the response of the flush call by `AbstractRunnable.run`.